File: python/cycle.py
from python.clock import Clock


class Cycle:
    def __init__(self,
                 duration_ticks,
                 on_up_fn,
                 on_down_fn,
                 is_one_shot):
        self.duration_ticks = duration_ticks
        self.on_up_fn = on_up_fn
        self.on_down_fn = on_down_fn
        self.is_one_shot = is_one_shot
        self.clock = Clock()
        self.iterations = 0

    def start(self):
        self.clock.start()

    def stop(self):
        self.clock.stop()

    def is_done(self):
        return self.is_one_shot and self.iterations > 0

    def next(self):
        if not self.clock.ticks is None and self.clock.ticks >= self.duration_ticks:
            self.iterations += 1
            if self.is_one_shot:
                self.stop()
                return
            else:
                # we're cycling; once self goes over the duration, we may as well reset back to 0
                self.clock.restart()

        # TODO: on the last step of the cycle, this will end up at a 0 point but _not_ rollover the next iteration
        self.clock.next()

        looping_elapsed_duration = self.clock.ticks % self.duration_ticks

        if self.on_down_fn is None:
            current_value = looping_elapsed_duration / self.duration_ticks
            self.on_up_fn(current_value)
        else:
            half_animation_loop_duration_ticks = self.duration_ticks / 2  # for up and down
            if looping_elapsed_duration < half_animation_loop_duration_ticks:
                # on our way up
                current_value = looping_elapsed_duration / half_animation_loop_duration_ticks
                self.on_up_fn(current_value)
            else:
                # on our way down
                current_value = (self.duration_ticks - looping_elapsed_duration) / half_animation_loop_duration_ticks
                self.on_down_fn(current_value)

    def jumpToDownCycle(self):
        if self.duration_ticks >= self.clock.ticks:
            # Don't "underun" the clock here: only jump down if it's meaningful to do so. This
            # shouldn't result in 0, otherwise we'll never hit a single iteration.
            # TODO: the problem with this logic is that it may never hit actual 0 value - just get close.
            self.clock.ticks = max(0, self.duration_ticks - self.clock.ticks)

    def restart(self, duration_ticks=None):
        if duration_ticks:
            self.duration_ticks = duration_ticks
        self.iterations = 0
        self.clock.restart()

    def is_at_zero_point(self):
        return self.clock.ticks is None or (self.clock.ticks % self.duration_ticks == 0)

File: src/Cycle.cpp
//
// Created by David Clay on 6/15/23.
//

#include "Cycle.h"
#include "Arduino.h"
#include <cmath>

Cycle::Cycle(int duration_ms, bool one_shot, CycleType cycle_type) {
    this->duration_ms = duration_ms;
    this->one_shot = one_shot;
    this->cycle_type = cycle_type;
    this->clock = new Clock();
    this->iterations = 0;
    this->release_phase = false;
    this->current_value = 0;
    this->last_iteration_seen = 0;
    this->at_zero_point = false;
}

void Cycle::start() {
    this->release_phase = false;
    // if the clock is on the falling side, set it to the equidistant rising side
    if (!this->isRising()) {
        unsigned long phase = this->clock->elapsed_ms % this->duration_ms;
        this->clock->shift_to(this->duration_ms - phase);
        // shift_to jumps elapsed_ms backwards relative to absolute iteration
        // count, so reset the crossing tracker to match the new phase.
        this->last_iteration_seen = 0;
    }
    this->clock->start();
}

void Cycle::stop() {
    this->clock->stop();
    this->current_value = 0;
    this->release_phase = false;
    this->last_iteration_seen = 0;
    this->at_zero_point = false;
}

void Cycle::reset() {
    // The `iterations` tracks whether this cycle has run or not.
    this->iterations = 0;
}

/**
 * Triggers the "release" phase of the cycle, dropping the
 * value down to its minimum which will then stop the cycle.
 */
void Cycle::release() {
    this->release_phase = true;
    if (this->isRising() && this->cycle_type == UP_DOWN_CYCLE) {
        // If we're on the rising side of the cycle, jump to the mirrored point
        // on the falling side at the same wave height, and continue to finish
        // the cycle. If we're on the falling side already, just let it finish.
        unsigned long phase = this->clock->elapsed_ms % this->duration_ms;
        this->clock->shift_to(this->duration_ms - phase);
        // shift_to jumps elapsed_ms backwards relative to absolute iteration
        // count, so reset the crossing tracker so the next boundary at
        // duration_ms is detected as a crossing.
        this->last_iteration_seen = 0;
    }
}

/**
 * Update this Cycle to the next iteration of its clock. Even if
 * a Cycle is stopped, it may have a release duration.
 */
void Cycle::update() {
    // always update the clock. If it's stopped, it's a no-op, otherwise ensures
    // the clock is up-to-date before doing any calculations.
    this->clock->update();
    this->at_zero_point = false;

    if (this->clock->isStopped()) {
        return;
    }

    // Iteration-crossing check: elapsed_ms may jump over the exact boundary,
    // so detect when (elapsed / duration) increments instead of (elapsed % duration == 0).
    unsigned long current_iteration = this->clock->elapsed_ms / this->duration_ms;
    if (current_iteration > this->last_iteration_seen) {
        this->iterations += static_cast<int>(current_iteration - this->last_iteration_seen);
        this->last_iteration_seen = current_iteration;
        this->at_zero_point = true;

        if (this->one_shot || this->release_phase) {
            this->stop();
            return;
        }
    } else if (this->clock->elapsed_ms == 0) {
        // first update after restart: original tick-based behavior also reported
        // ticks==0 as "at zero point" on the first update.
        this->at_zero_point = true;
    }

    unsigned long looping_elapsed = this->clock->elapsed_ms % this->duration_ms;
    if (this->cycle_type == UP_ONLY_CYCLE) {
        this->current_value = static_cast<float>(looping_elapsed) / static_cast<float>(this->duration_ms);
    } else {
        unsigned long half_duration = static_cast<unsigned long>(std::lround(this->duration_ms / 2.0));
        if (looping_elapsed <= half_duration) {
            this->current_value = static_cast<float>(looping_elapsed) / static_cast<float>(half_duration);
        } else {
            this->current_value = (static_cast<float>(this->duration_ms) - 1 - static_cast<float>(looping_elapsed)) / static_cast<float>(half_duration);
        }
    }
}

bool Cycle::isRising() const {
    if (this->cycle_type == UP_ONLY_CYCLE) {
        return true;
    }
    unsigned long relative = this->clock->elapsed_ms % this->duration_ms;
    return relative < static_cast<unsigned long>(this->duration_ms / 2);
}

void Cycle::restart() {
    this->restart(-1);
}

void Cycle::restart(int new_duration_ms) {
    if (new_duration_ms > -1) {
        this->duration_ms = new_duration_ms;
    }
    this->release_phase = false;
    this->iterations = 0;
    this->last_iteration_seen = 0;
    this->at_zero_point = false;
    this->clock->restart();
}

bool Cycle::isAtZeroPoint() const {
    return this->at_zero_point;
}

bool Cycle::isDone() const {
    return this->one_shot && this->iterations > 0;
}

bool Cycle::isStopped() const {
    return this->clock->isStopped();
}

bool Cycle::isRunning() const {
    return this->clock->running;
}
